src/AggregateFunctions/AggregateFunctionGroupSortedArray.h

```diff
@@ -0,0 +1,229 @@
+#pragma once
+
+#include <Columns/ColumnArray.h>
+#include <DataTypes/DataTypeArray.h>
+
+#include <AggregateFunctions/AggregateFunctionGroupSortedArrayData.h>
+#include <AggregateFunctions/IAggregateFunction.h>
+
+namespace DB
+{
+template <typename T, bool is_plain_column>
+inline T readItem(const IColumn * column, Arena * arena, size_t row)
+{
+    if constexpr (std::is_same_v<T, StringRef>)
+    {
+        if constexpr (is_plain_column)
+        {
+            StringRef str = column->getDataAt(row);
+            auto ptr = arena->alloc(str.size);
+            std::copy(str.data, str.data + str.size, ptr);
+            return StringRef(ptr, str.size);
+        }
+        else
+        {
+            const char * begin = nullptr;
+            return column->serializeValueIntoArena(row, *arena, begin);
+        }
+    }
+    else
+    {
+        return column->getUInt(row);
+    }
+}
+
+template <typename T>
+void getFirstNElements(const T * data, int num_elements, int threshold, size_t * results)
+{
+    for (int i = 0; i < threshold; i++)
+    {
+        results[i] = 0;
+    }
+
+    threshold = std::min(num_elements, threshold);
+    int current_max = 0;
+    int cur;
+    int z;
+    for (int i = 0; i < num_elements; i++)
+    {
+        //Starting from the highest values and we look for the immediately lower than the given one
+        for (cur = current_max; cur > 0 && (data[i] < data[results[cur - 1]]); cur--)
+            ;
+
+        if (cur < threshold)
+        {
+            //Move all the higher values 1 position to the right
+            for (z = current_max - 1; z >= cur; z--)
```

Comment:
It looks to me like an insertion sorting algorithm, or am I missing something?

Reply:
Yes, I need to get the first N elements. I made tests with other functions like std::nth_element() but I found this code more efficient.

Reply:
It would seem to me that this is only true up to a certain size, have you tested your implementation against std::nth_element and such (std::sort, std::partial_sort, std::partial_sort_copy, maybe even use std::priority_queue) for different Ns?

Because if it's true then we can make `if (x < threshold < y) { *use other implementation* }` and make everything even faster :)

Reply:
I created a repo to measure different methods. Maybe I missed something but I didn't find anything faster.

https://github.com/palegre-tiny/testfirstelements

Reply:
It was not my intention to code a new sorting function but I found it is much faster than other standard alternatives:

Current test results:
map 	  312132ns
sort   	1009483ns
custom1       19411ns
nth             146247ns

Reply:
I think I finally understood it, thanks.
I have added another selection algorimth for big array sizes, current code is 10x faster for short array sizes but for long sizes it is much more slower than std::nth_element (up to 100x). So I am using both, mine for threshold < 1000 and the standard one for bigger cases. 

I have also added performance tests for all (short, long) array sizes.